cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -475,1049 +207,821 @@ TEST_F(TestCast, CanCast) {
-  std::vector<int32_t> e1(is_valid.size(), 1);
-  for (size_t i = 0; i < v1.size(); ++i) {
-    if (i % 3 == 1) {
-      v1[i] = false;
-      e1[i] = 0;
-    }
-  }
-
-  CheckCase<BooleanType, Int32Type>(v1, is_valid, e1, options);
+TEST(Cast, FromBoolean) {
+  std::string vals = "[1, 0, null, 1, 0, 1, 1, null, 0, 0, 1]";
+  CheckCast(ArrayFromJSON(boolean(), vals), ArrayFromJSON(int32(), vals));
 }
 
-TEST_F(TestCast, ToBoolean) {
-  CastOptions options;
+TEST(Cast, ToBoolean) {
   for (auto type : kNumericTypes) {
-    CheckCaseJSON(type, boolean(), "[0, null, 127, 1, 0]",
-                  "[false, null, true, true, false]");
+    CheckCast(ArrayFromJSON(type, "[0, null, 127, 1, 0]"),
+              ArrayFromJSON(boolean(), "[false, null, true, true, false]"));
   }
 
   // Check negative numbers
-  CheckCaseJSON(int8(), boolean(), "[0, null, 127, -1, 0]",
-                "[false, null, true, true, false]");
-  CheckCaseJSON(float64(), boolean(), "[0, null, 127, -1, 0]",
-                "[false, null, true, true, false]");
+  for (auto type : {int8(), float64()}) {
+    CheckCast(ArrayFromJSON(type, "[0, null, 127, -1, 0]"),
+              ArrayFromJSON(boolean(), "[false, null, true, true, false]"));
+  }
 }
 
-TEST_F(TestCast, ToIntUpcast) {
-  CastOptions options;
-  options.allow_int_overflow = false;
-
+TEST(Cast, ToIntUpcast) {
   std::vector<bool> is_valid = {true, false, true, true, true};
 
   // int8 to int32
-  std::vector<int8_t> v1 = {0, 1, 127, -1, 0};
-  std::vector<int32_t> e1 = {0, 1, 127, -1, 0};
-  CheckCase<Int8Type, Int32Type>(v1, is_valid, e1, options);
-
-  // bool to int8
-  std::vector<bool> v2 = {false, true, false, true, true};
-  std::vector<int8_t> e2 = {0, 1, 0, 1, 1};
-  CheckCase<BooleanType, Int8Type>(v2, is_valid, e2, options);
+  CheckCast(ArrayFromJSON(int8(), "[0, null, 127, -1, 0]"),
+            ArrayFromJSON(int32(), "[0, null, 127, -1, 0]"));
 
   // uint8 to int16, no overflow/underrun
-  std::vector<uint8_t> v3 = {0, 100, 200, 255, 0};
-  std::vector<int16_t> e3 = {0, 100, 200, 255, 0};
-  CheckCase<UInt8Type, Int16Type>(v3, is_valid, e3, options);
+  CheckCast(ArrayFromJSON(uint8(), "[0, 100, 200, 255, 0]"),
+            ArrayFromJSON(int16(), "[0, 100, 200, 255, 0]"));
 }
 
-TEST_F(TestCast, OverflowInNullSlot) {
-  CastOptions options;
-  options.allow_int_overflow = false;
-
-  std::vector<bool> is_valid = {true, false, true, true, true};
-
-  std::vector<int32_t> v11 = {0, 70000, 2000, 1000, 0};
-  std::vector<int16_t> e11 = {0, 0, 2000, 1000, 0};
-
-  std::shared_ptr<Array> expected;
-  ArrayFromVector<Int16Type>(int16(), is_valid, e11, &expected);
-
-  auto buf = Buffer::Wrap(v11.data(), v11.size());
-  Int32Array tmp11(5, buf, expected->null_bitmap(), -1);
-
-  CheckPass(tmp11, *expected, int16(), options);
+TEST(Cast, OverflowInNullSlot) {
+  CheckCast(ArrayFromJSON(int32(), "[0, null, 2000, 1000, 0]"),
+            ArrayFromJSON(int16(), "[0, null, 2000, 1000, 0]"));
```

Comment:
Did you forget to call `MaskArrayWithNullsAt` in this test?

Reply:
Will do

Reply:
And now this test fails :-/